### ISSUE TITLE
Biomes transitions

### ISSUE THREAD

**Issue by kilbith** [minetest/minetest#2816]
Biomes transitions are currently too abrupt. They should be smooth such as in MC : 

![2t01g](https://cloud.githubusercontent.com/assets/7883281/8281513/f8ee1f02-18e8-11e5-8cb1-283e3d651907.jpg)

What I propose vaguely : 

![biomes_transition](https://cloud.githubusercontent.com/assets/7883281/8281746/445c5aa6-18eb-11e5-9ea8-10e877b40b08.jpg)

Also note : the engine ought not register new nodes for each colorization / overlaying.

@paramat What are your thoughts ?

**Comment by nerzhul** [comment]
:+1:

**Comment by rubenwardy** [comment]
I'd like this, but it must be custimisable by sub games.

**Comment by 4aiman** [comment]
After numerous "alike" feature requests have been deleted this one sprouted out and *finally* some devs "liked" the idea.

+:100:, since this has been desired for a couple of years now.

http://www.reddit.com/r/Minecraft/comments/1p2pgl/this_is_the_new_colour_system_texture_artists_are/

![Disco Biome](http://i4.minus.com/iewLk5W7nsZkX.jpg)

**Comment by paramat** [comment]
The idea is okay in isolation, but this would be slow, problematic and complex to put into practice.

I find Minecraft's smooth colour blend of grass too smooth, it's so gradual it looks continuous (maybe because it is a transition between two greens), which is perhaps the intention, but in a voxel world extreme smoothness looks out of place, textures should change in noticeable steps from voxel to voxel, as in the screenshot above in 4aiman's comment.

So a transition with a minimal number of steps looks okay, however, the problems:

Transitions are often between 3 to 5 biomes instead of 2, even with a limited number of steps the number of possible combinations becomes huge, it would be hundreds.
Texture artists would not want to design hundreds of textures. However if textures were automatically created they would not look good, for example imagine a transition between 2 radically different types of material/texture/colour, an auto-generated colour between may well be a crazy colour. An auto-generated texture which is a morph between the 2 would also often become a crazy texture.
However it was done it would be a big processing load added to mapgen, which needs to be fast and simple.

Also of course, a voxel world is inherently abrupt, to chase extreme smoothness is a ridiculous thing to do, voxel games embrace and welcome abruptness, it's part of their character.

So we have mgv6 biome blend which uses dither to create a transition, this is simple and suitable. I have just created a new noise-based biome blend for the current biome API which is better than both mgv6 and Minecraft biome blends.

The intention (of hmmmm) is to have multiple biome systems, he is planning a new biome system that may be more Minecraft-like, so you have more of a chance of influencing what ever biome blend he creates for that. The current mgv5/v7 biome API has been left mostly to me and i'm very happy with the new blobby noise biome blend, you can try it by using my biomesdev mod in mgv7.

**Comment by kilbith** [comment]
Thanks for this careful thought from expertise, @paramat

**Comment by 4aiman** [comment]
Wow, thanks, @paramat !

I'm not insisting or anything, but how about a different approach?

*E.g. How about such a feature as auto-generation of biomes?*

Say, there are 5 types of biomes.
Each has different node_top.
Let us say there are: sand, grass, snow, gravel, red_grass. (this is just an example)
Now, for transition to work, there should be a ```biome_colour``` field in the biome definition.
I guess that could be left to mod creators to fill.

Let us say the colours for the above biomes are: yellow, green, white, gray, red. 
if one put those near.... It would be horrible mess of an eye-bleed cause.

So, let us add yet another field: ```do_biome_blend```.
Set that to ```true``` and some part of the biome will be subject to change.

*Now on to that change:*

Look at the image below.
There are our biomes.
![before](https://cloud.githubusercontent.com/assets/2913210/8513026/4f73aca4-2364-11e5-8b3f-967ec2d11ce9.png)

Every biome have humidity and heat points. (If t doesn't, the "blending" won't be applies)

Now, if we just make the engine to loop through already registered biomes and make it possible to *add* some more technical transition biomes- that may do the trick!

It would be natural to add ```do_biome_blend``` to grassy biomes since the colour of grass may change due to be near to sand or gravel or snow (all those can be spilled all over the grass).

On the other hand it *would not* be natural to think that the nature will spread pieces of grass all over the snow biome or sand biome.

The ```change``` would imply the registration of a new node_top with all the properties of the original node_top but the texture. That may be got by ```[colorize```ing.

The decorations for a new biome may be copied or combined with the decorations of the nearby biome (may be subject to "obey" yet another field in the definnition).

So, when new biomes would be added the picture would look like this:
![after](https://cloud.githubusercontent.com/assets/2913210/8513027/52f9ea5a-2364-11e5-8de7-8d6b4d84b96d.png)
^ Note how red and green grass being mixed and the sand, snow and gravel fill some of the area of the technical biomes.


TL;DR

Right now it is possible to register some technical biomes and thus create blending.
But doing that by yourself is hell a lot of a trouble.
Why not make a blending which will increase the number of registered nodes (which isn't such a good idea) but won't abuse map generator in any way other than adding several new biomes.
I guess, more biomes won't be *so* bad in comparison with real-time texture colour calculations.

**Comment by Splizard** [comment]
Wouldn't the idea here to be texture blending between nodes with different textures?

![image](https://github.com/minetest/minetest/assets/2015791/a2eb6b5c-5e0f-4570-b69c-24d5b971cb4b)
